File: AdventOfCode2019/14/day14.js
const readInput = (file = "input") => {
  var fs = require("fs");
  var path = process.cwd();
  var buffer = fs.readFileSync(path + "\\14\\" + file);
  return buffer.toString();
};

const parseReactions = (str) => {
  const lines = str.split("\n").map((x) => x.trim());
  const rules = {};
  lines.forEach((line) => {
    const split = line.split(" => ");
    const output = split[1].split(" ");
    const inputs = split[0].split(", ").map((x) => x.split(" "));
    rules[output[1]] = {
      outputQuantity: +output[0],
      inputs: inputs.reduce((obj, cur) => ({ ...obj, [cur[1]]: +cur[0] }), {}),
    };
  });

  return rules;
};

const hasOnlyAvailableInput = (requirements, input) =>
  Object.keys(requirements).every((k) => k === input || requirements[k] <= 0);

const reduceByReaction = (reactions, requirements, input) => {
  let [requiredChemical, requiredQuantity] = Object.entries(requirements).find(
    ([chemical, amount]) => chemical !== input && amount > 0
  );

  // Reduce by reaction
  const reaction = reactions[requiredChemical];
  const applyCount = Math.ceil(requiredQuantity / reaction.outputQuantity);
  for (const inputChemical in reaction.inputs) {
    requirements[inputChemical] = requirements[inputChemical] || 0;
    requirements[inputChemical] += reaction.inputs[inputChemical] * applyCount;
  }

  const reactionOutput = reaction.outputQuantity * applyCount;
  requirements[requiredChemical] = requiredQuantity - reactionOutput;
  if (requirements[requiredChemical] === 0) {
    delete requirements[requiredChemical];
  }
};

const reduce = (reactions, requirements, inputChemical) => {
  while (!hasOnlyAvailableInput(requirements, inputChemical)) {
    reduceByReaction(reactions, requirements, inputChemical);
  }

  return requirements[inputChemical];
};

// Part 1
const input = readInput();
const reactions = parseReactions(input);
const orePerFuel = reduce(reactions, { FUEL: 1 }, "ORE");
console.log(orePerFuel);

// Part 2
// console.log(reduce(reactions, { FUEL: 3000000 }, "ORE"));
const maximizeFuel = (reactions, orePerFuel, inputOreQuantity) => {
  let low = Math.floor(inputOreQuantity / orePerFuel);
  let requiredOre = orePerFuel;
  let high = 3000000;
  while (low < high) {
    const mid = Math.ceil((high + low) / 2);
    requiredOre = reduce(reactions, { FUEL: mid }, "ORE");
    if (requiredOre > inputOreQuantity) {
      high = mid - 1;
    } else {
      low = mid;
    }
  }

  return low;
};

const max = maximizeFuel(reactions, orePerFuel, 1000000000000);
console.log(max);

File: AdventOfCode2019/14/Day14.hs
import Data.List (find)
import Data.List.Split (splitOn)
import Data.Maybe (fromJust, fromMaybe, isNothing)
import qualified Data.Map as M
import GHC.Integer (floatFromInteger)

type Chemical = (String, Integer) -- name, quantity

toChemical :: [Char] -> Chemical
toChemical s = let [quantity, name] = splitOn " " s in (name, read quantity :: Integer)

parseReaction :: [Char] -> (Chemical, [Chemical])
parseReaction s = (toChemical right, inputs)
  where
    [left, right] = splitOn " => " s
    inputs = map toChemical $ splitOn ", " left

parseReactions s =
  let xs = lines s in [parseReaction x | x <- xs]

findForReplace :: [Chemical] -> Maybe Chemical
findForReplace = find (\(n, q) -> n /= "ORE" && q > 0)

reduce' :: [(Chemical, [Chemical])] -> [Chemical] -> Integer
reduce' reactions requirements
  | isNothing replace = snd $ fromJust ore
  | otherwise = reduce' reactions newRequirements
  where
    replace = findForReplace requirements
    ore = find (\(n, _) -> n == "ORE") requirements
    Just (chemical, quantity) = replace
    Just ((rChemical, rQuantity), rInput) = find (\((n, _), _) -> n == chemical) reactions
    (quot, rem) = quotRem quantity rQuantity
    applyCount = if rem > 0 then quot + 1 else quot
    reactionOutput = rQuantity * applyCount
    toAdd = map (\(n, q) -> (n, q * applyCount)) rInput
    n1 = foldl (\acc (n, q) -> M.insertWith (+) n q acc) (M.fromList requirements) toAdd
    n2 = M.insertWith (flip (-)) chemical reactionOutput n1 
    newRequirements = M.toList n2

maximizeFuel :: [(Chemical, [Chemical])] -> t -> Integer -> Float -> Float -> Float
maximizeFuel reactions orePerFuel oreLimit low high
  | low > high = low
  | otherwise = maximizeFuel reactions orePerFuel oreLimit low1 high1
  where
    mid = ceiling $ (low + high) / 2
    midF = fromIntegral mid
    requiredOre = reduce' reactions [("FUEL", mid)]
    [low1, high1] = if requiredOre > oreLimit then [low, midF - 1] else [midF, high]

part1 reactions = reduce' reactions [("FUEL", 1)]

part2 reactions orePerFuel = maximizeFuel reactions orePerFuel oreLimit (fromIntegral low) (fromIntegral high)
  where
    oreLimit = 1000000000000
    low = oreLimit `div` orePerFuel
    high = 3000000

main = do
  content <- readFile "input"
  let reactions = parseReactions content
  let p1 = part1 reactions 
  print p1
  print $ part2 reactions p1